### PR TITLE
PE-878: 2022-03-11 Executive (goerli)

### PR DESCRIPTION
**Pull request by hexonaut**
# Description

# Contribution Checklist

- [ ] PR title starts with `(PE-<TICKET_NUMBER>)`
- [ ] Code approved
- [ ] Tests approved
- [ ] CI Tests pass

# Checklist

- [ ] Every contract variable/method declared as public/external private/internal
- [ ] Consider if this PR needs the `officeHours` modifier override
- [ ] Verify expiration (`30 days` unless otherwise specified)
- [ ] Verify hash in the description matches [here](https://emn178.github.io/online-tools/keccak_256.html)
- [ ] Validate all addresses used are in Goerli changelog or known
- [ ] Notify any external teams affected by the spell so they have the opportunity to review
- [ ] Deploy spell to Goerli `ETH_GAS="XXX" ETH_GAS_PRICE="YYY" make deploy`
- [ ] Ensure contract is verified on `Goerli` etherscan
- [ ] Change test to use Goerli spell address and deploy timestamp
- [ ] Cast spell on Goerli `make spell="0x-deployed-spell-address" cast-spell`
- [ ] Run `make archive-spell` or `make date="YYYY-MM-DD" archive-spell` to make an archive directory and copy `Goerli-DssSpell.sol`, `Goerli-DssSpell.t.sol`, `Goerli-DssSpell.t.base.sol`, and `Goerli-DssSpellCollateralOnboarding.sol`
- [ ] `squash and merge` this PR

### CODE REVIEW

**Comment by gbalabasquer** [comment]
Minor typo in the PR name, should be 878

File: src/Goerli-DssSpell.sol [hunk 1]
```diff
@@ -21,24 +21,39 @@ pragma solidity 0.6.12;
...
+        DssExecLib.setValue(MCD_FLAP, "ttl", FlapAbstract(MCD_FLAP_OLD).ttl());
+        DssExecLib.setValue(MCD_FLAP, "tau", FlapAbstract(MCD_FLAP_OLD).tau());
+        DssExecLib.setValue(MCD_FLAP, "lid", 150_000 * RAD);
+        DssExecLib.deauthorize(MCD_FLAP_OLD, MCD_VOW);
+        DssExecLib.authorize(MCD_FLAP, MCD_VOW);
+        DssExecLib.setContract(MCD_VOW, "flapper", MCD_FLAP);
+
+        // Changelog updates
+        DssExecLib.setChangelogAddress("MCD_FLAP", MCD_FLAP);
+        DssExecLib.setChangelogVersion("1.10.1");
```
Comment (julienmartinlevrai) [inline src/Goerli-DssSpell.sol:56]: As I suggested in the chat, I think adding a pre-release tag to all Görli versions makes sense. That way it is possible to update the pre-release tag if we need to re-deploy. Also, if you think about it, Görli deployments *are* pre-releases.

This allows us to ensure compatibility between chain versions (latest Görli pre-release == Mainnet release) without worrying too much about things like maintaining the order in which things get deployed in every chain.

```suggestion
        DssExecLib.setChangelogVersion("1.10.1-alpha");
```

Comment (gbalabasquer) [reply src/Goerli-DssSpell.sol:56]: I agree we need to revisit the Goerli procedure of versioning, however we should stick to the actual one until a group decision is taken. I think that should be outside of any specific PR.

File: src/Goerli-DssSpell.t.sol [hunk 2]
```diff
@@ -385,4 +381,50 @@ contract DssSpellTest is GoerliDssSpellTestBase {
         }
         assertEq(expectedHash, actualHash);
     }
+
+    function setFlaps() internal {
+        vote(address(spell));
+        spell.schedule();
+        hevm.warp(spell.nextCastTime());
+        spell.cast();
+        assertTrue(spell.done());
```
Comment (naszam) [inline src/Goerli-DssSpell.t.sol:390]:
```suggestion
        vote(address(spell));
        scheduleWaitAndCast(address(spell));
        assertTrue(spell.done());
```

File: src/Goerli-DssSpell.t.base.sol [hunk 3]
```diff
@@ -34,6 +34,11 @@ interface DirectDepositLike is GemJoinAbstract {
     function king() external view returns (address);
 }
 
+interface FlapLike is FlapAbstract {
+    function fill() external view returns (uint256);
+    function lid() external view returns (uint256);
```
Comment (naszam) [inline src/Goerli-DssSpell.t.base.sol:39]: Missing `lid` in System Values and range check. (https://github.com/makerdao/spells-goerli/pull/52/commits/f397e0ba44737c308ca6a79ae7b70c57f222acc4#diff-66a33be4e15ec87add1b15eb9fe8e292c82f68ffe873e045f2bda0588b2fea1c)
Also, there is a base test refactoring (https://github.com/makerdao/spells-goerli/pull/52/commits/8bf495f3837a39b820f1e536e0631ca4c31dc43f) that would require to address conflicts once merged on master.

File: src/Goerli-DssSpell.sol [hunk 4]
```diff
@@ -21,24 +21,39 @@ pragma solidity 0.6.12;
...
 
 import { DssSpellCollateralOnboardingAction } from "./Goerli-DssSpellCollateralOnboarding.sol";
 
 contract DssSpellAction is DssAction, DssSpellCollateralOnboardingAction {
     // Provides a descriptive tag for bot consumption
     string public constant override description = "Goerli Spell";
 
+    uint256 public constant RAD = 10**45;
+
+    address public constant MCD_FLAP = 0x015bEd3a7EBbB0Be03A35E0572E8a7B0BA2AA0fB;
```
Comment (naszam) [inline src/Goerli-DssSpell.sol:35]: nit: consider removing `public` getter

**Review by naszam — COMMENTED** [review]
Checked:

- Office Hours Off
- 30 Days Expiry (weekly)
- Pragma Ok
- Math Ok (`RAD`)
- Rate Limited Flap (verified, constructor args ok (vat, gem), wards ok (pause proxy relied, deployer denied), no opts, license ok, matches dss [flap-rate-limit](https://github.com/makerdao/dss/blob/flap-rate-limit/src/flap.sol), github)
- ChainLog Changes Ok (Flap, `v1.10.1`)
- `addresses_mainnet.sol` Matches
- DssExecLib v0.0.9
- Spell Actions Look Good
- Tests Look Good
- Tests PASS
```haskell
dapp-test: rpc block: latest
Running 15 tests for src/Goerli-DssSpell.t.sol:DssSpellTest
[PASS] testFail_notScheduled() (gas: 7139)
[PASS] test_auth_in_sources() (gas: 3250690)
[PASS] testFail_new_flapper_exeed_limit() (gas: 1007164)
[PASS] test_use_eta() (gas: 331274)
[PASS] test_new_flapper() (gas: 1009655)
[PASS] testOnTime() (gas: 557982)
[PASS] test_nextCastTime() (gas: 332214)
[PASS] testFailTooEarly() (gas: 3640)
[PASS] test_bytecode_matches() (gas: 1503788)
[PASS] test_auth() (gas: 2473407)
[PASS] testSpellIsCast_GENERAL() (gas: 8018731)
[PASS] testFailWrongDay() (gas: 3662)
[PASS] testCastCost() (gas: 559549)
[PASS] testNewChainlogValues() (gas: 564026)
[PASS] testFailTooLate() (gas: 3639)
```
Notes: Missing `lid` in SystemValues and `lid` range check

LGTM

File: src/Goerli-DssSpell.t.sol (hunk 2 above)
Comment (hexonaut) [reply src/Goerli-DssSpell.t.sol]: done

File: src/Goerli-DssSpell.t.base.sol (hunk 3 above)
Comment (hexonaut) [reply src/Goerli-DssSpell.t.base.sol:39]: done

File: src/Goerli-DssSpell.sol (hunk 4 above)
Comment (hexonaut) [reply src/Goerli-DssSpell.sol]: done

File: src/Goerli-DssSpell.sol (hunk 1 above)
Comment (hexonaut) [reply src/Goerli-DssSpell.sol:56]: going to leave this for this spell

File: src/Goerli-DssSpell.sol [hunk 5]
```diff
@@ -29,35 +30,30 @@ contract DssSpellAction is DssAction, DssSpellCollateralOnboardingAction {
...
-
-        // Decrease DIRECT-AAVEV2-DAI Target Borrow Rate (bar) from 3.5% to 2.85%.
-        // DssExecLib.setD3MTargetInterestRate(DssExecLib.getChangelogAddress("MCD_JOIN_DIRECT_AAVEV2_DAI"), 285); // 2.85%
+        // Replace Flapper with rate limit one
+        address MCD_VOW = DssExecLib.vow();
+        address MCD_FLAP_OLD = DssExecLib.flap();
+        DssExecLib.setValue(MCD_FLAP, "beg", FlapAbstract(MCD_FLAP_OLD).beg());
+        DssExecLib.setValue(MCD_FLAP, "ttl", FlapAbstract(MCD_FLAP_OLD).ttl());
+        DssExecLib.setValue(MCD_FLAP, "tau", FlapAbstract(MCD_FLAP_OLD).tau());
+        DssExecLib.setValue(MCD_FLAP, "lid", 150_000 * RAD);
```
Comment (naszam) [inline src/Goerli-DssSpell.sol:49]: I think `lid` was set to 300K on the other pr (https://github.com/makerdao/spells-goerli/commit/f397e0ba44737c308ca6a79ae7b70c57f222acc4#diff-66a33be4e15ec87add1b15eb9fe8e292c82f68ffe873e045f2bda0588b2fea1c)

File: src/Goerli-DssSpell.sol [hunk 6]
```diff
@@ -29,35 +30,30 @@ contract DssSpellAction is DssAction, DssSpellCollateralOnboardingAction {
...
-        // Increase WSTETH-A Autoline (gap) from 20 million DAI to 30 million DAI.
-        DssExecLib.setIlkAutoLineParameters("WSTETH-A", 300 * MILLION, 30 * MILLION, 6 hours);
-
-        // Increase DIRECT-AAVEV2-DAI AutoLine (line) from 220 million DAI to 300 million DAI.
-        // Increase DIRECT-AAVEV2-DAI AutoLine (gap) from 50 million DAI to 65 million DAI.
-        // DssExecLib.setIlkAutoLineParameters("DIRECT-AAVEV2-DAI", 300 * MILLION, 65 * MILLION, 12 hours);
-
-        // Decrease DIRECT-AAVEV2-DAI Target Borrow Rate (bar) from 3.5% to 2.85%.
-        // DssExecLib.setD3MTargetInterestRate(DssExecLib.getChangelogAddress("MCD_JOIN_DIRECT_AAVEV2_DAI"), 285); // 2.85%
+        // Replace Flapper with rate limit one
```
Comment (julienmartinlevrai) [inline src/Goerli-DssSpell.sol:43]:
```suggestion
        // Replace Flapper with rate limit one
        // https://vote.makerdao.com/polling/Qmdd4Pg7
```

File: src/Goerli-DssSpell.sol [hunk 7]
```diff
@@ -29,35 +30,30 @@ contract DssSpellAction is DssAction, DssSpellCollateralOnboardingAction {
...
+        address MCD_VOW = DssExecLib.vow();
+        address MCD_FLAP_OLD = DssExecLib.flap();
+        DssExecLib.setValue(MCD_FLAP, "beg", FlapAbstract(MCD_FLAP_OLD).beg());
+        DssExecLib.setValue(MCD_FLAP, "ttl", FlapAbstract(MCD_FLAP_OLD).ttl());
+        DssExecLib.setValue(MCD_FLAP, "tau", FlapAbstract(MCD_FLAP_OLD).tau());
+        DssExecLib.setValue(MCD_FLAP, "lid", 150_000 * RAD);
+        DssExecLib.deauthorize(MCD_FLAP_OLD, MCD_VOW);
```
Comment (julienmartinlevrai) [inline src/Goerli-DssSpell.sol:50]: note that there is no need to do this

**Comment by hexonaut** [comment]
Fixed

File: src/Goerli-DssSpell.t.sol [hunk 8]
```diff
@@ -357,4 +356,48 @@ contract DssSpellTest is GoerliDssSpellTestBase {
...
+    }
+
+    function testFail_new_flapper_exeed_limit() public {
+        setFlaps();
+
+        vow.flap();
+        vow.flap();
+        vow.flap();
+        vow.flap();
+        vow.flap();
```
Comment (julienmartinlevrai) [inline src/Goerli-DssSpell.t.sol:400]: I think this way of testing the failure ensures that the test is failing for the reason we expect it to fail.

```suggestion
        test_new_flapper()
```

**Review by julienmartinlevrai — APPROVED** [review]
### General spell correctness

* RAD correctly defined

#### Tests

* deployed spell empty

### Flapper replacement

* FlapAbstract correctly imported. It corresponds to the older version, but the newer is not needed
* New Flapper contract address added
* Vow and older Flapper correctly assigned
* The three old params are correctly copied from the old Flapper
* The new `lid` param is correctly set to 150K
* The vow is correctly deauthorized in the old Flapper (although there is no need)
* The vow is correctly authorized in the new Flapper
* The new Flapper is correctly filed in the Vow
* No other contracts need to be updated in our core infrastructure
* The Chainlog is updated correctly
* The Chainlog version’s patch component is bumped

#### Deployment of new Flapper contract

* Contract [deployed](https://goerli.etherscan.io/address/0x015bEd3a7EBbB0Be03A35E0572E8a7B0BA2AA0fB)
* Contract verified on Etherscan
* Verified code matches repository code
* The only ward is the Pause Proxy
* Constructor arguments are correct
   - Vat address is correct
   - Gem address is MKR ERC-20 contract

#### Chainlog tests

* Flapper address correctly updated in `addresses_goerli.sol`
* Chainlog address check is correct
* Chainlog version check is correct

#### Flapper tests

* `flap_lid` variable added to Config contract
* `flap_lid` correctly set to 150K
* Forced creation of 1B dai for the vow correct
* Flapper address checks correct
* `fill` checks correct
* limit exceeded check correct

### Tests PASS

```Haskell
dapp-test: rpc block: latest
Running 14 tests for src/Goerli-DssSpell.t.sol:DssSpellTest
[PASS] testFail_notScheduled() (gas: 7139)
[PASS] testFail_new_flapper_exeed_limit() (gas: 1016840)
[PASS] test_use_eta() (gas: 331874)
[PASS] test_new_flapper() (gas: 1010017)
[PASS] testOnTime() (gas: 558537)
[PASS] test_nextCastTime() (gas: 332819)
[PASS] testFailTooEarly() (gas: 3550)
[PASS] test_bytecode_matches() (gas: 1470164)
[PASS] testSpellIsCast_GENERAL() (gas: 7988952)
[PASS] testFailWrongDay() (gas: 3528)
[PASS] testCastCost() (gas: 560060)
[PASS] testNewChainlogValues() (gas: 564559)
[PASS] testFailTooLate() (gas: 3527)
[PASS] testRemoveChainlogValues() (gas: 572533)
```

**Comment by hexonaut** [comment]
Updated from @julienmartinlevrai's review. Leaving the de-authorize just to be safe.

**Comment by naszam** [comment]
`test_auth` and `test_auth_in_sources` tests can be enabled (by setting the functions to `public`) as there is a new ChainLog entry (Flap).
(see https://github.com/makerdao/spells-goerli/pull/52/commits/afb86b670c8961f6db24d39c568b14be63a47862#diff-5244fd19ae9af910cfb59c8bbf07d1459638b1e089ae1bc01141417f6016ab6a)

**Review by naszam — APPROVED** [review]
Checked:

- Spell Code Changes Look Good (`public` getters, poll link)
- Test Changes Look Good 
	- `deployed_spell` unset
	-  Flap specific test optimization
	- add `lid` to SystemValues in `config` and range checks in base tests
	- `scheduleWaitAndCast` pattern
	- disabled `testRemoveChainlogValues` via `private` trick
	- enabled `test_auth` and `test_auth_sources`
- Test PASS
```haskell
dapp-test: rpc block: latest
Running 15 tests for src/Goerli-DssSpell.t.sol:DssSpellTest
[PASS] testFail_notScheduled() (gas: 7139)
[PASS] test_auth_in_sources() (gas: 3855541)
[PASS] testFail_new_flapper_exeed_limit() (gas: 1016773)
[PASS] test_use_eta() (gas: 331896)
[PASS] test_new_flapper() (gas: 1010039)
[PASS] testOnTime() (gas: 558559)
[PASS] test_nextCastTime() (gas: 332841)
[PASS] testFailTooEarly() (gas: 3572)
[PASS] test_bytecode_matches() (gas: 1470186)
[PASS] test_auth() (gas: 10037131)
[PASS] testSpellIsCast_GENERAL() (gas: 7988907)
[PASS] testFailWrongDay() (gas: 3572)
[PASS] testCastCost() (gas: 560104)
[PASS] testNewChainlogValues() (gas: 564581)
[PASS] testFailTooLate() (gas: 3549)
```

LGTDeploy

**Review by julienmartinlevrai — APPROVED** [review]
* deployed spell address added
* contract verified on Etherscan
* Etherscan code matches locally flattened code
* DssExecLib address matches https://github.com/makerdao/dss-exec-lib/blob/master/README.md
* timestamp matches deployment block
* tests pass

```haskell
dapp-test: rpc block: latest
Running 15 tests for src/Goerli-DssSpell.t.sol:DssSpellTest
[PASS] testFail_notScheduled() (gas: 7139)
[PASS] test_auth_in_sources() (gas: 3822819)
[PASS] testFail_new_flapper_exeed_limit() (gas: 984051)
[PASS] test_use_eta() (gas: 331896)
[PASS] test_new_flapper() (gas: 977317)
[PASS] testOnTime() (gas: 525837)
[PASS] test_nextCastTime() (gas: 332841)
[PASS] testFailTooEarly() (gas: 3572)
[PASS] test_bytecode_matches() (gas: 1470186)
[PASS] test_auth() (gas: 10004409)
[PASS] testSpellIsCast_GENERAL() (gas: 7956185)
[PASS] testFailWrongDay() (gas: 3572)
[PASS] testCastCost() (gas: 527382)
[PASS] testNewChainlogValues() (gas: 531859)
[PASS] testFailTooLate() (gas: 3549)
```

LGTM

**Review by julienmartinlevrai — APPROVED** [review]
Archived folder is identical to `src/`, good to merge

**Review by naszam — APPROVED** [review]
Checked:

* Deployed Spell Verified
	* Optimization Enabled: No with 200 run
	* Other Settings: default evmVersion, GNU GPLv3 license
	* DssExecLib v0.0.9
* Deploy Timestamp Matches
* Deployed Spell Code Matches GitHub ([diffcheck](https://www.diffchecker.com/G9NPh9dp))
* Archive Looks Good (diffed)
* Test PASS (with deployed spell address)
```haskell
dapp-test: rpc block: 6479545
Running 15 tests for src/Goerli-DssSpell.t.sol:DssSpellTest
[PASS] testFail_notScheduled() (gas: 7139)
[PASS] test_auth_in_sources() (gas: 3855541)
[PASS] testFail_new_flapper_exeed_limit() (gas: 1016773)
[PASS] test_use_eta() (gas: 331896)
[PASS] test_new_flapper() (gas: 1010039)
[PASS] testOnTime() (gas: 558559)
[PASS] test_nextCastTime() (gas: 332841)
[PASS] testFailTooEarly() (gas: 3572)
[PASS] test_bytecode_matches() (gas: 1470186)
[PASS] test_auth() (gas: 10037131)
[PASS] testSpellIsCast_GENERAL() (gas: 7988907)
[PASS] testFailWrongDay() (gas: 3572)
[PASS] testCastCost() (gas: 560104)
[PASS] testNewChainlogValues() (gas: 564581)
[PASS] testFailTooLate() (gas: 3549)
```
Notes: we should reach consensus on goerli archive date pattern as currently there is a bit of asynchrony with the cast dates.

LGTMerge

**Pull request merged**